src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,711 @@
+        """Converts samples into Qibolab results format.
+
+        Args:
+            ro_pulse_list (list): List of readout pulse sequences.
+            samples (dict): Samples generated by self.get_samples.
+            append_to_shape (list): Specifies additional dimensions for the shape of the results. Defaults to empty list.
+            execution_parameters (`qibolab.ExecutionParameters`): Parameters (nshots,
+                                                        relaxation_time,
+                                                        fast_reset,
+                                                        acquisition_type,
+                                                        averaging_mode)
+
+        Returns:
+            dict: Qibolab results for AcquisitionType.DISCRIMINATION.
+
+        Raises:
+            TypeError: If execution_parameters.acquisition_type is AcquisitionType.RAW or AcquisitionType.INTEGRATION.
+        """
+        shape = [execution_parameters.nshots] + append_to_shape
+        results = {}
+        for ro_pulse in ro_pulse_list:
+            if execution_parameters.acquisition_type is AcquisitionType.DISCRIMINATION:
+                values = np.array(samples[ro_pulse.qubit]).reshape(shape)
+                processed_values = SampleResults(values)
+
+                if execution_parameters.averaging_mode is AveragingMode.SINGLESHOT:
+                    pass
+                elif execution_parameters.averaging_mode is AveragingMode.CYCLIC:
+                    processed_values = (
+                        processed_values.average
+                    )  # generaetes AveragedSampleResults
+
+            elif execution_parameters.acquisition_type is AcquisitionType.RAW:
+                raise TypeError("Emulator does not support raw measurement")
+            elif execution_parameters.acquisition_type is AcquisitionType.INTEGRATION:
+                raise TypeError("Emulator does not support integrated measurement")
+
+            results[ro_pulse.qubit] = results[ro_pulse.serial] = processed_values
+        return results
+
+    def run_pulse_simulation(
+        self,
+        sequence: PulseSequence,
+        instant_measurement: bool = True,
+    ) -> tuple[np.ndarray, list]:
+        """Simulates the input pulse sequence.
+
+        Args:
+            sequence (`qibolab.pulses.PulseSequence`): Pulse sequece to simulate.
+            instant_measurement (bool): Collapses readout pulses to duration of 1 if True. Defaults to True.
+
+        Returns:
+            tuple: Reduced density matrix of final state specified by readout qubits as well as the list of readout qubits in, both in little endian order.
+        """
+        # reduces measurement time to 1 dt to save simulation time
+        if instant_measurement:
+            for i in range(len(sequence)):
+                if type(sequence[i]) is ReadoutPulse:
+                    sequence[i].duration = 1
```

Comment:
This can be simplified as
```suggestion
            for pulse in sequence:
                if type(pulse) is ReadoutPulse:
                    pulse.duration = 1
```
However, there is another issue with this because it is secretly mutating the sequence that the user is passing. If the user uses `sequence` again later somewhere else in the code they may get unexpected results.

Reply:
I see, so I should instead create a copy of the modified sequence instead?

Reply:
Why a copy? The only thing changing in the proposal is the iteration (if you have a sequence, it is better to iterate its elements, instead of iterating indices and accessing the elements - think about linked lists)

Reply:
My point on creating a copy was with reference to @stavros11 's comment that "it is secretly mutating the sequence that the user is passing. If the user uses `sequence` again later somewhere else in the code they may get unexpected results." I'm agreeable with the iteration simplification.

Reply:
Ah yes, sorry, I missed that part. That's also sensible, but you will need something like `copy.deepcopy(sequence)`, because `sequence.copy()` would be shallow, thus you'd be mutating user pulses anyhow (and so also the `sequence`, indirectly, in the very same of no `.copy()`).

Reply:
So it's an option that can be not exercised, depending on the user. The main reason is that the readout pulse simulation/model is non-existent yet. The implementation at the moment is to mimic sample generation from `AcquisitionType.DISCRIMINATION` by sampling from the (diagonal) of the final density matrix of the full system. Therefore there is no need to wait for the additional (idle) simulation time coming from the readout pulses. This is also why currently only `AcquisitionType.DISCRIMINATION` is supported.

Reply:
Maybe it could be worth to sample at different times during the measurement window. Despite missing the model for the readout pulse, you know that the discrimination is taking a value of I and Q acquired over an extended interval. Thus, even just the bare evolution (without ro pulse) may have a non-trivial effect.

(Of course, having access to the whole waveform would be better, but this would also require some more modelling, while throwing the collapse at random in the window is "for free", up to the cost of the extra evolution).

Reply:
Diagonal in the local Hamiltonian eigenstates, correct?

Reply:
Explicitly, let $\rho$ be the density matrix of the full system at the end of the pulse simulation (prior to readout).
1) $\rho_{RO} = tr_{RO'}(\rho)$ is the reduced density matrix for all the qubits with readout pulses (RO) obtained by tracing out the rest of the system (RO').
2) $P_{RO} = Diag(\rho_{RO})$ is then the probability distribution of all the outcomes (computational basis) of the measured qubits. I sample from $P_{RO}$ and convert the outcome into a bitstring (or the generalised version of it when more than 2 levels are simulated for the readout qubits).

Reply:
Fixed in e91a1532c49b64d10b01afbc35f11140a1605f2e